### DISCUSSION TITLE
How to connect multiple, separate physical servers

### DISCUSSION THREAD

**Discussion by ConfusedMerlin** [DeviceFarmer/stf#617] [Q&A]
Greetings, 

I have three servers (one master, two nodes), which each have their own pool of devices. Two of them (the nodes) are set up to report their devices to the first one (the master). To connect the devices from the node-servers to the master-server, their provider-service was set up to have its --connect-push and --connect-sub aiming for the the tri-proxy service of the master-server. 

All of these servers have docker containers, each one containing one of the necessary stf services, with all parameters defined inside the docker-compose.yml.  One difference between node and master is that the nodes provider point to the triproxy of the master, while the master points to itself.  Also, the nginx server config of the nodes redirects http requests to the master. Right now, the nodes have all services up too, even if they seem to be unnecessary. 

This setup seems to work, as the web GUI of the master shows all devices attached to the other two nodes and its own as well. 
 It seems to be compliant to the deployment.md, regarding what must be located where together and what must speak to which (device or appliance side) triproxy service. 

But is that the correct way to set this up?

**Comment by denis99999** [comment]
@ConfusedMerlin ,
The fact that the GUI indicates that all devices are connected does not validate your configuration because, for example, taking control of a device can reveal a gray screen or another problem like an instability etc. If so, you should describe it using the "problem model" and you should also provide your configuration files (nginx, docker-compose, etc.) for analysis.

Also, I recommend that you clearly separate what you call "master" and "node" servers to dedicated machines, i.e. reserving the device provider function to node machines only, without any other services, with the exception of ADB of course.

**Reply by ConfusedMerlin** [reply]
Okay, so the minimum services/units for a "node" type server (only providing more devices to the total farm) are the provider and their adb?

I will set up that somewhere to find out what fails afterwards. If nothing blows up, I will try it at the environment where the Grey Screens appear. I bet you are correct about this being a config connection problem somewhere.

**Reply by denis99999** [reply]
> Okay, so the minimum services/units for a "node" type server (only providing more devices to the total farm) are the provider and their adb?

Yes, this is why you don't need a powerfull machine for the device provider unit which is stateless, for example a simple Raspberry Pi3 is enough for ~6 devices connected through a stable USB 2.0 hub.
